tgui/packages/tgui/interfaces/PaperSheet.tsx

```diff
@@ -0,0 +1,881 @@
+/**
+ * @license MIT
+ */
+
+import { classes } from 'common/react';
+import { useBackend, useLocalState } from '../backend';
+import { Box, Button, Flex, Section, TextArea } from '../components';
+import { Window } from '../layouts';
+import { sanitizeText } from '../sanitize';
+import { marked } from 'marked';
+import { Component, createRef, RefObject } from 'inferno';
+import { clamp } from 'common/math';
+
+const Z_INDEX_STAMP = 1;
+const Z_INDEX_STAMP_PREVIEW = 2;
+
+const TEXTAREA_INPUT_HEIGHT = 200;
+
+type PaperContext = {
+  // ui_static_data
+  user_name: string;
+  raw_text_input?: PaperInput[];
+  raw_field_input?: FieldInput[];
+  raw_stamp_input?: StampInput[];
+  max_length: number;
+  max_input_field_length: number;
+  paper_color: string;
+  paper_name: string;
+  default_pen_font: string;
+  default_pen_color: string;
+  signature_font: string;
+
+  // ui_data
+  held_item_details?: WritingImplement;
+};
+
+type PaperInput = {
+  raw_text: string;
+  font?: string;
+  color?: string;
+  bold?: boolean;
+};
+
+type StampInput = {
+  class: string;
+  x: number;
+  y: number;
+  rotation: number;
+};
+
+type FieldInput = {
+  field_index: string;
+  field_data: PaperInput;
+  is_signature: boolean;
+};
+
+type WritingImplement = {
+  interaction_mode: InteractionType;
+  font?: string;
+  color?: string;
+  use_bold?: boolean;
+  stamp_icon_state?: string;
+  stamp_class?: string;
+};
+
+type PaperSheetStamperState = {
+  x: number;
+  y: number;
+  rotation: number;
+  yOffset: number;
+};
+
+type PaperSheetStamperProps = {
+  scrollableRef: RefObject<HTMLDivElement>;
+};
+
+type FieldCreationReturn = {
+  nextCounter: number;
+  text: string;
+};
+
+type StampPosition = {
+  x: number;
+  y: number;
+  rotation: number;
+  yOffset: number;
+};
+
+enum InteractionType {
+  reading = 0,
+  writing = 1,
+  stamping = 2,
+}
+
+type PreviewViewProps = {
+  scrollableRef: RefObject<HTMLDivElement>;
+  handleOnScroll: (this: GlobalEventHandlers, ev: Event) => any;
+  textArea: string;
+};
+
+const canEdit = (heldItemDetails?: WritingImplement): boolean => {
+  if (!heldItemDetails) {
+    return false;
+  }
+
+  return heldItemDetails.interaction_mode === InteractionType.writing;
+};
+
+// Regex that finds [____] fields.
+const fieldRegex: RegExp = /\[((?:_+))\]/gi;
+
+// Handles the ghost stamp when attempting to stamp paper sheets.
+class PaperSheetStamper extends Component<PaperSheetStamperProps> {
+  style: null;
+  handleMouseMove: (this: Document, ev: MouseEvent) => any;
+  handleMouseClick: (this: Document, ev: MouseEvent) => any;
+  state: PaperSheetStamperState = { x: 0, y: 0, rotation: 0, yOffset: 0 };
+  scrollableRef: RefObject<HTMLDivElement>;
+
+  constructor(props, context) {
+    super(props, context);
+
+    this.style = null;
+    this.scrollableRef = props.scrollableRef;
+
+    this.handleMouseMove = (e) => {
```

Comment:
you could implement this in the same way as pauseEvent below, for consistency and to get rid of a redundant declaration at the top.

Reply:
This and the mouse click event handler functions have been moved into the body of the class instead of the constructor, decalarations removed.